### PR TITLE
(#5) Definition Display

### LINKED ISSUE #5: Story: Definition Display
## As a content editor, I would like to display the definition of a term at the top of search results in order to provide more detailed information to users if their search term matches a term in NCI’s dictionary.

> **ESTIMATE** 5

### Acceptance Criteria

```gherkin
Scenario: English Dictionary Definition Display
    Given the user navigates to "/?swKeyword=metastasis"
     Then the page title is "NCI Search Results"
      And page subtitle "Results for: metastasis" appears below the page title
      And definition box appears with title "Definition:"
      And the word "metastasis" appears in the definition box, with the audio icon and pronunciation
      And the definition "The spread of cancer cells from the place where they first formed to another part of the body." appears in the definition box (the first sentence of the definition)
      And link to the definition page with text "More information on dictionary page" and href "https://www.cancer.gov/publications/dictionaries/cancer-terms/def/46710" in the definition box
      And a button to toggle the full definition appears in the definition box labelled "Show full definition"
     When user clicks on the full definition toggle button in the definition box
     Then full definition toggle button text turns to "Hide full definition"
    Given user is viewing the second page of results for "metastasis"
      And page subtitle "Results for: metastasis" appears below the page title
     Then the definition box no longer appears on the page
```

```gherkin
Scenario: English Dictionary Definition Display - no more information link (see notes)
    Given the user navigates to "/?swKeyword=tumor"
     Then the page title is "NCI Search Results"
      And page subtitle "Results for: tumor" appears below the page title
      And definition box appears with title "Definition:"
      And the word "metastasis" appears in the definition box, with the audio icon and pronunciation
      And the definition "An abnormal mass of tissue that results when cells divide more than they should or do not die when they should." appears in the definition box
      And link to the definition page with text "More information on dictionary page" does not display
      And a button to toggle the full definition appears in the definition box labelled "Show full definition"
     When user clicks on the full definition toggle button in the definition box
     Then full definition toggle button text turns to "Hide full definition"
```

```gherkin
  Scenario: English Dictionary Definition Display - no show/hide full definition button
    Given the user navigates to "/?swKeyword=caloric intake"
     Then the page title is "NCI Search Results"
      And page subtitle "Results for: caloric intake" appears below the page title
      And definition box appears with title "Definition:"
      And the word "caloric intake" appears in the definition box, with the audio icon and pronunciation
      And the definition "Refers to the number of calories (energy content) consumed." appears in the definition box
      And link to the definition page with text "More information on dictionary page" does not display
      And button to toggle the full definition in the definition box labelled "Show full definition" does not display
```

```gherkin
Scenario: Spanish Dictionary Definition Display
    Given the user navigates to "/?swKeyword=cáncer de mama"
     Then the page title is "Resultados"
      And page subtitle "Resultados para: cáncer de mama" appears below the page title
      And definition box appears with title "Definición:"
      And the word "cáncer de mama" appears in the definition box with the audio icon
      And the definition "Cáncer que se forma en los tejidos del seno (mama)." appears in the definition box
      And link to the definition page with text "Más información" and href "https://www.cancer.gov/espanol/publicaciones/diccionario/def/444971" in the definition box
      And a button to toggle the full definition appears in the definition box labelled "Mostrar toda la definición"
     When user clicks on the full definition toggle button in the definition box
     Then full definition toggle button text turns to "Ocultar toda la definición"
    Given user is viewing the second page of results for "cáncer de mama"
      And page subtitle "Resultados para: cáncer de mama" appears below the page title
     Then the definition box no longer appears on the page
```

```gherkin
Scenario: Spanish Dictionary Definition Display - no more information link
    Given the user navigates to "/?swKeyword=tumor"
     Then the page title is "Resultados"
      And page subtitle "Resultados para: tumor" appears below the page title
      And definition box appears with title "Definición:"
      And the word "tumor" appears in the definition box with the audio icon
      And the definition "Masa anormal de tejido que aparece cuando las células se multiplican más de lo debido o no se destruyen en el momento apropiado." appears in the definition box
      And link to the definition page with text "Más información" does not display
      And a button to toggle the full definition appears in the definition box labelled "Mostrar toda la definición"
     When user clicks on the full definition toggle button in the definition box
     Then full definition toggle button text turns to "Ocultar toda la definición"
```

```gherkin
  Scenario: Spanish Dictionary Definition Display - no show/hide full definition button
    Given the user navigates to "/?swKeyword=cáncer de la vagina en estadio lll"
     Then the page title is "Resultados"
      And page subtitle "Resultados para: cáncer de la vagina en estadio lll" appears below the page title
      And definition box appears with title "Definición:"
      And the word "cáncer de la vagina en estadio lll" appears in the definition box with the audio icon
      And the definition "El cáncer se diseminó hasta la pared de la pelvis." appears in the definition box
      And link to the definition page with text "Más información" and href "https://www.cancer.gov/espanol/publicaciones/diccionario/def/535780" in the definition box
      And button to toggle the full definition in the definition box labelled "Mostrar toda la definición" does not display
```

### Resources:
* Feature: JIRA Ticket #
* Mockups: *Here goes a URL to or the name of the mockup(s) in inVision*
* Web Page: *Here goes a URL to an existing web page*

### Notes

* Search term and definition must be exact match.
   * Spanish accents are not required.
* For `And a link to the definition page appears below the first sentence with the words "More information on dictionary page"` this link should be in the form, `<glossary_url>/def/<prettyurl_or_id>` where:
  * glossary_url is something like `https://colo.cancer.gov/publications/dictionaries/cancer-terms`
  * prettyurl_or_id is the pretty url name if available, otherwise use the id
* More information only appears if there are media or related resources for the term then the link should display 

---

### Prerequisites
- *required dependency for identifying the problem, scope and then solution for this issue*
- [ ] Short Spike Description - Issue #9999

### Tasks
- [x] Create `Definition` component
- [x] Fetch results from API
- [x] Apply styling
- [x] Add unit tests
- [x] Add integration tests

### Technical Notes
- More notes, or technical nuances/direction

### PR DESCRIPTION
**Pull request by seyilonge-nci**
Closes #5

* Added useURLQuery hook
* Added `Definition` display to `Home`
* Fetch results from API
* `Definition` toggle feature and applied styling
* Added unit tests
* Integration tests (WIP...)
* Integration tests complete
* Hide "More Information.." link based on API response
* Hide definition box when results page is not the first page
* Updated setupProxy file for dictionary call
* Updated mock responses
* Added responseInterceptor for dictionary response
* Updated unit tests
* Commented out 'More Information' link step in integration tests since current response always has 'relatedResources' empty
* Set doneLoading to true when no keyword has been provided
* Apply case-insensitivity to 'swKeyword' get from url
* Only show full definition toggle button when definition has more than one sentence
* Added styling for chevron to show/hide full definition button
* Modified api call adding requestedFields for required data
* Added media to condition for displaying 'More info..' link
* Updated mock data, Restored 'More info..' link step in integration test
* Removed responseInterceptor
* Updated font styling on Show/Hide full definition button text
* Added additional integration test for show/hide full definition for one sentent definitions
* Updated styling based on UX review
* Updates after rebase to reconcile state changes and api fetch calls
* Moved title to Home, and applied styling to h2
* Added meta tag for noindex directive to head
* Added spanish mock file for video
* Updated tracking when search payload doesn't exist

### CODE REVIEW

**Comment by seyilonge-nci** [comment]
@VictoriaSunNIH @lburack @blilianyu @mworrest 
Please review:
https://react-app-dev.cancer.gov/sitewide-search-app/ticket__5-definition-display
https://react-app-dev.cancer.gov/sitewide-search-app/ticket__5-definition-display?cfg=0&swkeyword=cancer
https://react-app-dev.cancer.gov/sitewide-search-app/ticket__5-definition-display?cfg=1&swkeyword=c%C3%A1ncer

**Comment by blilianyu** [comment]
@seyilonge-nci  "Show full definition" should be Noto Sans and #2E2E2E. It's displaying as Arial and #000000.

**Comment by blilianyu** [comment]
@seyilonge-nci H2 should be 20px not 23px.

**Comment by mworrest** [comment]
Please decrease padding above title "Definition" to 16px.

**Comment by seyilonge-nci** [comment]
@blilianyu and @mworrest Updated! Please could you review again.

**Comment by mworrest** [comment]
Looking good--thanks!

File: src/components/molecules/definition/definition.scss [hunk 1]
```diff
@@ -0,0 +1,59 @@
+#dictionary-definition {
```
Comment (arcepaul) [inline src/components/molecules/definition/definition.scss:1]: Try not to use ids in css(makes it harder to overcome), it's already a .definition so why not use that as the landmark?

File: src/components/molecules/search-results-list/search-results-list.jsx [hunk 2]
```diff
@@ -43,11 +44,17 @@ const SearchResultsList = ({
 		);
 	});
 
+	const renderHelmet = () => {
+		return (
```
Comment (arcepaul) [inline src/components/molecules/search-results-list/search-results-list.jsx:48]: why is this being handled by the search results list?  This is something that should be handled by the view.

File: support/src/setupProxy.js [hunk 3]
```diff
@@ -179,18 +155,18 @@ const mockNoResultsAPI = (err) => {
...
-	);
-
 	app.use(
 		'/api/sitewidesearch/v1/Search/:collection/:language/:term',
 		getSearchResults
 	);
 
 	app.use('/api/BestBets/:collection/:language/:term', getBestBetsResults);
 
+	app.use(
```
Comment (arcepaul) [inline support/src/setupProxy.js:165]: Why hardcode what dictionary this will use?  Remember that this library could be used by other people for other purposes and applications.  I would switch back to the configurable path

File: src/components/molecules/definition/definition.scss (hunk 1 above)
Comment (seyilonge-nci) [reply src/components/molecules/definition/definition.scss:1]: I actually absolutely refrain from using ids to style, but I was also looking to match current prod as much as possible to avoid issues. Although we no longer have `best-bet-definition` id felt a name swap would be easier that's why I went with this. Would gladly remove this.

File: support/src/setupProxy.js (hunk 3 above)
Comment (seyilonge-nci) [reply support/src/setupProxy.js:165]: 👍

File: src/components/molecules/search-results-list/search-results-list.jsx (hunk 2 above)
Comment (seyilonge-nci) [reply src/components/molecules/search-results-list/search-results-list.jsx:48]: Will switch.

File: support/src/setupProxy.js (hunk 3 above)
Comment (seyilonge-nci) [reply support/src/setupProxy.js:165]: Done!

File: src/components/molecules/search-results-list/search-results-list.jsx (hunk 2 above)
Comment (seyilonge-nci) [reply src/components/molecules/search-results-list/search-results-list.jsx]: Done!

File: src/components/molecules/definition/definition.scss (hunk 1 above)
Comment (seyilonge-nci) [reply src/components/molecules/definition/definition.scss]: Done!
